arangod/Replication2/StateMachines/Document/DocumentLeaderState.cpp

```diff
@@ -105,6 +105,14 @@ auto DocumentLeaderState::recoverEntries(std::unique_ptr<EntryIterator> ptr)
 
     while (auto entry = ptr->next()) {
       auto doc = entry->second;
+      if (doc.operation == OperationType::kCreateShard ||
+          doc.operation == OperationType::kDropShard) {
+        // TODO
+        // If the maintenance sees that shards are missing, it is going to
+        // create them on the leader.
+        // Perhaps we should check that they're already created.
+        continue;
+      }
```

Comment:
If we do not create the shard immediately, how do we apply log entries for those shards (which might already exist in the log)

Reply:
True! This is a problem I've been thinking about yesterday. I was guessing what would happen in the maintenance at the point where a replicated state becomes leader. For the maintenance to create shards on the leader, it would have to know that the local replicated state is the leader of the shard sheaf in which the missing shard belongs to. This implies that the leader state is instantiated before the maintenance tries to create the shard, which I don't know if it's always the case.
I think a possible solution would be to let the leader create the shard in recovery, and when the maintenance runs, it would instruct the leader to create the shard again, but it would only do so unless it wasn't already created yet.

Reply:
I created [CINFRA-656](https://arangodb.atlassian.net/browse/CINFRA-656) to tackle this problem. I suggest to let the core create the shards in this PR.